pkg/standalone-utils/contracts/TokenPairRegistry.sol

```diff
@@ -0,0 +1,242 @@
+import { InputHelpers } from "@balancer-labs/v3-solidity-utils/contracts/helpers/InputHelpers.sol";
+
+import { OwnableAuthentication } from "./OwnableAuthentication.sol";
+
+contract TokenPairRegistry is ITokenPairRegistry, OwnableAuthentication {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    mapping(bytes32 pairId => IBatchRouter.SwapPathStep[][] paths) internal _pairsToPaths;
+
+    constructor(IVault vault, address initialOwner) OwnableAuthentication(vault, initialOwner) {
+        // solhint-disable-previous-line no-empty-blocks
+    }
+
+    function getPathAt(
+        address tokenA,
+        address tokenB,
+        uint256 index
+    ) external view returns (IBatchRouter.SwapPathStep[] memory) {
+        bytes32 tokenId = _getTokenPairId(tokenA, tokenB);
+        return _pairsToPaths[tokenId][index];
+    }
+
+    function getPathCount(address tokenA, address tokenB) external view returns (uint256) {
+        bytes32 tokenId = _getTokenPairId(tokenA, tokenB);
+        return _pairsToPaths[tokenId].length;
+    }
+
+    function getPaths(address tokenA, address tokenB) external view returns (IBatchRouter.SwapPathStep[][] memory) {
+        bytes32 tokenId = _getTokenPairId(tokenA, tokenB);
+        return _pairsToPaths[tokenId];
+    }
+
+    function addPath(address tokenIn, IBatchRouter.SwapPathStep[] memory steps) external authenticate {
+        address tokenOut = address(steps[steps.length - 1].tokenOut);
+        bytes32 tokenId = _getTokenPairId(tokenIn, tokenOut);
+
+        address stepTokenIn = tokenIn;
+        for (uint256 i = 0; i < steps.length; ++i) {
+            IBatchRouter.SwapPathStep memory step = steps[i];
+            if (step.isBuffer) {
+                _checkBufferStep(step.pool, stepTokenIn, address(step.tokenOut));
+            } else {
+                _checkPoolStep(step.pool, stepTokenIn, address(step.tokenOut));
+            }
+
+            // Update token in for the next iteration
+            stepTokenIn = address(step.tokenOut);
+        }
+
+        IBatchRouter.SwapPathStep[][] storage paths = _pairsToPaths[tokenId];
+        paths.push();
+        for (uint256 i = 0; i < steps.length; ++i) {
+            paths[paths.length - 1].push(steps[i]);
+        }
+        emit PathAdded(tokenIn, tokenOut, _pairsToPaths[tokenId].length);
+    }
+
+    function removePathAtIndex(address tokenIn, address tokenOut, uint256 index) external authenticate {
+        bytes32 tokenId = _getTokenPairId(tokenIn, tokenOut);
+        IBatchRouter.SwapPathStep[][] storage paths = _pairsToPaths[tokenId];
+        uint256 pathsLength = paths.length;
+
+        if (index >= pathsLength) {
+            revert IndexOutOfBounds();
+        }
+
+        if (pathsLength > 1 && index != pathsLength - 1) {
+            paths[index] = paths[pathsLength - 1];
+        }
+        // pop() can be used to clear dynamic arrays: it deletes every element and removes the inner array entirely.
+        paths.pop();
```

Comment:
Do we need to validate that the path still makes sense after the removal, like we do when adding a path? In that case, could factor out the "path validation" code and call it both places.